sfxclient/multitokensink.go

```diff
@@ -57,62 +112,25 @@ func newAtomicBool() *atomicBool {
 
 // worker manages a pipeline for emitting metrics
 type worker struct {
-	lock         *sync.Mutex       // lock to control concurrent access to the worker
-	input        chan *msg         // channel for inputing datapoints into a worker
-	maxBuffer    int               // maxBuffer is the maximum size of the buffer for data
-	errorHandler func(error) error // error handler for handling error emitting datapoints
-	dataBuffered int64             // atomic Int64 that maintains count of data in the sink
-	sink         *HTTPSink         // sink is an HTTPSink for emitting datapoints to Signal Fx
-	done         chan bool         // done is a boolean channel for signaling that the worker has stopped
-}
-
-// emits a series of datapoints
-func (w *worker) emit(token string, datapoints []*datapoint.Datapoint, events []*event.Event) {
-	// set the token on the HTTPSink
-	w.sink.AuthToken = token
-
-	if len(datapoints) > 0 {
-		// emit datapoints and handle any errors
-		if err := w.sink.AddDatapoints(context.Background(), datapoints); err != nil {
-			_ = w.errorHandler(err)
-		}
-		// account for the emitted datapoints
-		_ = atomic.AddInt64(&w.dataBuffered, int64(len(datapoints)*-1))
-	}
-	if len(events) > 0 {
-		// emit datapoints and handle any errors
-		if err := w.sink.AddEvents(context.Background(), events); err != nil {
-			_ = w.errorHandler(err)
-		}
-
-		// account for the emitted events
-		_ = atomic.AddInt64(&w.dataBuffered, int64(len(events)*-1))
-	}
-}
-
-// Stop will stop the pipeline
-func (w *worker) Stop(timeout time.Duration) bool {
-	for {
-		select {
-		case <-time.After(timeout): // timeout and generate error message indicating what may be dropped
-			return false
-		case w.input <- stopMsg:
-			// stop datapoint emission
-		case <-w.done:
-			defer close(w.done)
-			return true
-		}
-	}
+	lock                *sync.Mutex       // lock to control concurrent access to the worker
+	errorHandler        func(error) error // error handler for handling error emitting datapoints
+	sink                *HTTPSink         // sink is an HTTPSink for emitting datapoints to Signal Fx
+	closing             chan bool         // channel to signal that the worker is stopping
+	stopped             *atomicBool       // indicates that the worker has completely shutdown
+	invalidStatusRegexp *regexp.Regexp    // regex to identify an http error from the sink
+	statusCodeRegexp    *regexp.Regexp    // regex to identify the status code returned in an error from the sink
```

Comment:
whoah, no. do we really need regexes up in this? A) they're potentially really slow B) you own the AddDatapoints (and all downstream code).  If you can't get what you want from sfxclient then maybe changes need to be made there, but this is just straight up "ick".